### ISSUE TITLE
Completed tasks from a failing worker are retried

### ISSUE THREAD

**Issue by john-jam** [PrefectHQ/prefect#12970]
When a Dask worker fails after some time, all the completed tasks that ran on this worker are retried on another Dask worker.

This prevents using long-running tasks on spot instances in cloud environment like GCP where a worker is expected to fail at some point and the retry of an already completed task is not wanted.

I asked how we could avoid retrying already completed tasks on [an issue](https://github.com/dask/distributed/issues/7870#issuecomment-1570512357) in the `dask/distributed` repository. As indicated in the issue, this behavior is the default behavior of a Dask client since in the DaskTaskRunner, a reference for each task future is held in the `self._dask_futures` dictionary.

In order to not retry completed tasks, one could use the `as_completed` class as described in [this response](https://github.com/dask/distributed/issues/7870#issuecomment-1570512357).

# Expectation / Proposal

The completed tasks from a failing worker are not retried on another worker.

# Traceback / Example

Here's a minimal complete verifiable example based on the one I created in the linked `dask/distributed` issue but with the help of prefect:

I ran a dask scheduler and 2 workers into separate terminals with the following commands:

```bash
# Scheduler
$ dask scheduler

# Worker 1
$ dask worker 127.0.0.1:8786 --nworkers 1 --nthreads 1

# Worker 2
$ dask worker 127.0.0.1:8786 --nworkers 1 --nthreads 1
```

And here's the python code used to test:

```python
import time

from prefect import flow, task
from prefect_dask import DaskTaskRunner


@task
def dummy_task(task_id: int):
    print(f"Task {task_id}")
    time.sleep(5)
    return f"result_{task_id}"


@flow(task_runner=DaskTaskRunner(address="127.0.0.1:8786"))
def dummy_flow():
    for i in range(15):
        dummy_task.submit(task_id=i)


if __name__ == '__main__':
    dummy_flow()
```

I stopped the `worker 2` after it processed successfully 5 tasks and here are the logs:

- Scheduler
```bash
2023-06-02 00:15:32,292 - distributed.scheduler - INFO - -----------------------------------------------
2023-06-02 00:15:32,295 - distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
2023-06-02 00:15:32,297 - distributed.scheduler - INFO - State start
2023-06-02 00:15:32,299 - distributed.scheduler - INFO - -----------------------------------------------
2023-06-02 00:15:32,299 - distributed.scheduler - INFO -   Scheduler at:    tcp://10.10.80.57:8786
2023-06-02 00:15:32,299 - distributed.scheduler - INFO -   dashboard at:                     :8787
2023-06-02 00:15:35,464 - distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:43723', status: init, memory: 0, processing: 0>
2023-06-02 00:15:35,466 - distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:43723
2023-06-02 00:15:35,466 - distributed.core - INFO - Starting established connection to tcp://127.0.0.1:60534
2023-06-02 00:15:36,977 - distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:42631', status: init, memory: 0, processing: 0>
2023-06-02 00:15:36,977 - distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:42631
2023-06-02 00:15:36,977 - distributed.core - INFO - Starting established connection to tcp://127.0.0.1:60548
2023-06-02 00:15:42,012 - distributed.scheduler - INFO - Receive client connection: Client-2c62ff88-008f-11ee-87bd-13b66c7c6c40
2023-06-02 00:15:42,012 - distributed.core - INFO - Starting established connection to tcp://127.0.0.1:60578
2023-06-02 00:16:10,997 - distributed.core - INFO - Connection to tcp://127.0.0.1:60548 has been closed.
2023-06-02 00:16:10,997 - distributed.scheduler - INFO - Remove worker <WorkerState 'tcp://127.0.0.1:42631', status: running, memory: 5, processing: 2>
2023-06-02 00:16:10,997 - distributed.core - INFO - Removing comms to tcp://127.0.0.1:42631
```

- Worker 1
```bash
2023-06-02 00:15:35,194 - distributed.nanny - INFO -         Start Nanny at: 'tcp://127.0.0.1:36081'
2023-06-02 00:15:35,455 - distributed.diskutils - INFO - Found stale lock file and directory '/tmp/dask-worker-space/worker-v7igrctq', purging
2023-06-02 00:15:35,461 - distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:43723
2023-06-02 00:15:35,461 - distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:43723
2023-06-02 00:15:35,461 - distributed.worker - INFO -          dashboard at:            127.0.0.1:46685
2023-06-02 00:15:35,461 - distributed.worker - INFO - Waiting to connect to:       tcp://127.0.0.1:8786
2023-06-02 00:15:35,461 - distributed.worker - INFO - -------------------------------------------------
2023-06-02 00:15:35,461 - distributed.worker - INFO -               Threads:                          1
2023-06-02 00:15:35,461 - distributed.worker - INFO -                Memory:                   1.92 GiB
2023-06-02 00:15:35,461 - distributed.worker - INFO -       Local Directory: /tmp/dask-worker-space/worker-c3mncyw6
2023-06-02 00:15:35,461 - distributed.worker - INFO - -------------------------------------------------
2023-06-02 00:15:35,466 - distributed.worker - INFO -         Registered to:       tcp://127.0.0.1:8786
2023-06-02 00:15:35,466 - distributed.worker - INFO - -------------------------------------------------
2023-06-02 00:15:35,467 - distributed.core - INFO - Starting established connection to tcp://127.0.0.1:8786
Task 10
00:15:49.506 | INFO    | Task run 'dummy_task-10' - Finished in state Completed()
Task 4
00:15:54.667 | INFO    | Task run 'dummy_task-4' - Finished in state Completed()
Task 3
00:15:59.822 | INFO    | Task run 'dummy_task-3' - Finished in state Completed()
Task 2
00:16:04.955 | INFO    | Task run 'dummy_task-2' - Finished in state Completed()
Task 12
00:16:10.080 | INFO    | Task run 'dummy_task-12' - Finished in state Completed()
Task 9
00:16:15.205 | INFO    | Task run 'dummy_task-9' - Finished in state Completed()
Task 11
00:16:20.330 | INFO    | Task run 'dummy_task-11' - Finished in state Completed()
Task 13
00:16:25.484 | INFO    | Task run 'dummy_task-13' - Finished in state Completed()
Task 1
00:16:30.619 | INFO    | Task run 'dummy_task-1' - Finished in state Completed()
Task 0
00:16:35.735 | INFO    | Task run 'dummy_task-0' - Finished in state Completed()
Task 6
00:16:40.899 | INFO    | Task run 'dummy_task-6' - Finished in state Completed()
Task 7
00:16:46.043 | INFO    | Task run 'dummy_task-7' - Finished in state Completed()
Task 5
00:16:51.170 | INFO    | Task run 'dummy_task-5' - Finished in state Completed()
Task 8
00:16:56.301 | INFO    | Task run 'dummy_task-8' - Finished in state Completed()
Task 14
00:17:01.414 | INFO    | Task run 'dummy_task-14' - Finished in state Completed()
```

- Worker 2
```bash
2023-06-02 00:15:36,699 - distributed.nanny - INFO -         Start Nanny at: 'tcp://127.0.0.1:44573'
2023-06-02 00:15:36,974 - distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:42631
2023-06-02 00:15:36,974 - distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:42631
2023-06-02 00:15:36,974 - distributed.worker - INFO -          dashboard at:            127.0.0.1:35739
2023-06-02 00:15:36,974 - distributed.worker - INFO - Waiting to connect to:       tcp://127.0.0.1:8786
2023-06-02 00:15:36,974 - distributed.worker - INFO - -------------------------------------------------
2023-06-02 00:15:36,974 - distributed.worker - INFO -               Threads:                          1
2023-06-02 00:15:36,974 - distributed.worker - INFO -                Memory:                   1.92 GiB
2023-06-02 00:15:36,974 - distributed.worker - INFO -       Local Directory: /tmp/dask-worker-space/worker-kdyqin2d
2023-06-02 00:15:36,974 - distributed.worker - INFO - -------------------------------------------------
2023-06-02 00:15:36,978 - distributed.worker - INFO -         Registered to:       tcp://127.0.0.1:8786
2023-06-02 00:15:36,978 - distributed.worker - INFO - -------------------------------------------------
2023-06-02 00:15:36,978 - distributed.core - INFO - Starting established connection to tcp://127.0.0.1:8786
Task 0
00:15:49.485 | INFO    | Task run 'dummy_task-0' - Finished in state Completed()
Task 1
00:15:54.650 | INFO    | Task run 'dummy_task-1' - Finished in state Completed()
Task 13
00:15:59.800 | INFO    | Task run 'dummy_task-13' - Finished in state Completed()
Task 6
00:16:04.935 | INFO    | Task run 'dummy_task-6' - Finished in state Completed()
Task 7
00:16:10.067 | INFO    | Task run 'dummy_task-7' - Finished in state Completed()
Task 5
^C2023-06-02 00:16:10,985 - distributed._signals - INFO - Received signal SIGINT (2)
2023-06-02 00:16:10,986 - distributed.nanny - INFO - Closing Nanny at 'tcp://127.0.0.1:44573'. Reason: nanny-close
2023-06-02 00:16:10,986 - distributed.nanny - INFO - Nanny asking worker to close. Reason: nanny-close
00:16:10.986 | ERROR   | Task run 'dummy_task-5' - Crash detected! Execution was cancelled by the runtime environment.
2023-06-02 00:16:10,997 - distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:42631. Reason: worker-handle-scheduler-connection-broken
00:16:10.997 | INFO    | distributed.worker - Stopping worker at tcp://127.0.0.1:42631. Reason: worker-handle-scheduler-connection-broken
^CProcess Dask Worker process (from Nanny):
2023-06-02 00:16:11,150 - distributed.dask_worker - INFO - End worker
```

- Python code test
```bash
00:15:42.005 | INFO    | prefect.engine - Created flow run 'sincere-snail' for flow 'dummy-flow'
00:15:42.007 | INFO    | prefect.task_runner.dask - Connecting to an existing Dask cluster at 127.0.0.1:8786
00:15:42.013 | INFO    | prefect.task_runner.dask - The Dask dashboard is available at http://127.0.0.1:8787/status
00:15:42.193 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-0' for task 'dummy_task'
00:15:42.443 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-0' for execution.
00:15:42.451 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-10' for task 'dummy_task'
00:15:42.455 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-10' for execution.
00:15:42.505 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-1' for task 'dummy_task'
00:15:42.508 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-1' for execution.
00:15:42.518 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-4' for task 'dummy_task'
00:15:42.522 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-4' for execution.
00:15:42.532 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-13' for task 'dummy_task'
00:15:42.536 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-13' for execution.
00:15:42.607 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-3' for task 'dummy_task'
00:15:42.611 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-3' for execution.
00:15:42.621 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-6' for task 'dummy_task'
00:15:42.624 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-6' for execution.
00:15:42.708 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-2' for task 'dummy_task'
00:15:42.711 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-2' for execution.
00:15:42.721 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-7' for task 'dummy_task'
00:15:42.725 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-7' for execution.
00:15:42.734 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-12' for task 'dummy_task'
00:15:42.738 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-12' for execution.
00:15:42.810 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-5' for task 'dummy_task'
00:15:42.814 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-5' for execution.
00:15:42.823 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-9' for task 'dummy_task'
00:15:42.827 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-9' for execution.
00:15:42.920 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-8' for task 'dummy_task'
00:15:42.928 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-8' for execution.
00:15:43.029 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-11' for task 'dummy_task'
00:15:43.036 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-11' for execution.
00:15:43.127 | INFO    | Flow run 'sincere-snail' - Created task run 'dummy_task-14' for task 'dummy_task'
00:15:43.131 | INFO    | Flow run 'sincere-snail' - Submitted task run 'dummy_task-14' for execution.
00:17:01.478 | INFO    | Flow run 'sincere-snail' - Finished in state Completed('All states completed.')
```

The `worker 2` processed successfully the tasks `0`, `1`, `13`, `6`, and `7` but we can see that those tasks are re-ran on `worker 1` (the retry of the task `5` is expected since I closed the `worker 2` when it executed this task).

- [x] I would like to [help contribute](https://PrefectHQ.github.io/prefect-dask/#contributing) a pull request to resolve this!

**Comment by cicdw** [comment]
This can be configured by using task cache policies; see documentation here: https://docs.prefect.io/3.0/develop/task-caching

**Comment by john-jam** [comment]
@cicdw Hi :wave: 
I tried to reproduce this issue with prefect v3 and it seems resolved (the tasks of a failing worker are no longer re-executed). But I'm wondering if the "fix" is here to stay because now, the `DaskTaskRunner`'s `__enter__` method is called on every tasks:
```
04:55:00.686 | INFO    | prefect.task_runner.dask - Connecting to an existing Dask cluster at tcp://localhost:8786
04:55:00.711 | INFO    | Task run 'run_long_task-70f' - Task 193 started (delay: 1s)
04:55:01.712 | INFO    | Task run 'run_long_task-70f' - Task 193 finished (delay: 1s)
04:55:01.714 | INFO    | Task run 'run_long_task-70f' - Finished in state Completed()
04:55:02.186 | INFO    | prefect.task_runner.dask - Connecting to an existing Dask cluster at tcp://localhost:8786
04:55:02.213 | INFO    | Task run 'run_long_task-3b5' - Task 195 started (delay: 1s)
04:55:03.215 | INFO    | Task run 'run_long_task-3b5' - Task 195 finished (delay: 1s)
04:55:03.221 | INFO    | Task run 'run_long_task-3b5' - Finished in state Completed()
04:55:03.668 | INFO    | prefect.task_runner.dask - Connecting to an existing Dask cluster at tcp://localhost:8786
04:55:03.695 | INFO    | Task run 'run_long_task-85e' - Task 197 started (delay: 1s)
04:55:04.696 | INFO    | Task run 'run_long_task-85e' - Task 197 finished (delay: 1s)
04:55:04.699 | INFO    | Task run 'run_long_task-85e' - Finished in state Completed()
04:55:05.143 | INFO    | prefect.task_runner.dask - Connecting to an existing Dask cluster at tcp://localhost:8786
04:55:05.168 | INFO    | Task run 'run_long_task-8a9' - Task 198 started (delay: 1s)
04:55:06.169 | INFO    | Task run 'run_long_task-8a9' - Task 198 finished (delay: 1s)
04:55:06.171 | INFO    | Task run 'run_long_task-8a9' - Finished in state Completed()
```

Is that a new intended behavior for the DaskTaskRunner to enter a new context on the dask worker for each tasks? Works for me but wondering if this will change in a future prefect version.

**Comment by cicdw** [comment]
Hmmm no I don't believe that's intended; will need to investigate why that's happening

**Comment by john-jam** [comment]
> Hmmm no I don't believe that's intended; will need to investigate why that's happening

Thanks for you answer! But then, if the previous behavior gets reverted, I've got some concerns the bug defined [here](https://github.com/PrefectHQ/prefect/issues/12970#issue-2266033784) will show up again.
It's a really annoying one that basically blocks a prefect flow in an infinite loop (workers keeps failing and making other workers fail and so on).

**Comment by cicdw** [comment]
@john-jam ok I looked into this, and I was wrong: this _is_ expected behavior. Here's what's going on:

- the task is submitted to a worker
- once the task run is picked up by a worker, it hydrates the prefect task run context (which includes initiating the `DaskTaskRunner` and reconnecting to the scheduler - this is important behavior for users who interact directly with the dask client on the runner and / or submit nested tasks)

**Comment by john-jam** [comment]
@cicdw Awesome new! Thanks for looking into this. We can keep this issue as closed then 🙇‍♂ 

I got a slightly related question (please let me know if I should ask it somewhere else): On an OSS prefect server deployment in k8s with a flow with a lot of parallel tasks (>1000), when observing the tasks status and flow status from the UI, the flow status gets updated to `COMPLETED` (and the flow run pod terminates successfully) but some of the tasks status are still `PENDING` for a while and gets updated to `COMPLETED` only after some time. Does prefect v3 have a background task running somewhere to update tasks status after a flow is completed? It does not happen in prefect v2.

**Comment by cicdw** [comment]
@john-jam yup that is correct - flow run state updates occur synchronously, but task run state updates occur through the new events backend, which is a background service. This was one of the key architectural changes that gave 3.0 a runtime performance boost, at the cost of the UI "buffer time" that you've noticed. We've been increasingly shifting our focus back towards performance issues lately, so hopefully we can continue to improve on the observed latency of state updates!

**Comment by john-jam** [comment]
Hello @cicdw and @desertaxle 👋 

Sorry to comment on a closed issue (please let me know if I should create a new issue) but I experience the same behavior in the latest version of prefect (`prefect==3.3.1/prefect-dask==0.3.4`).

### Reproducible Example

**Flow code:**

```python
import time
from prefect import flow, task
from prefect_dask import DaskTaskRunner


@task
def dummy_task(task_id: int):
    print(f"Task {task_id}")
    time.sleep(5)
    return f"result_{task_id}"


@flow(task_runner=DaskTaskRunner(address="127.0.0.1:8786"))
def dummy_flow():
    # Run 10 dummy tasks with dask
    futures = dummy_task.map(task_id=[i for i in range(10)])
    futures.result()


if __name__ == '__main__':
    dummy_flow()
```

**Prefect Flow run logs:**

```bash
16:34:49.291 | INFO    | Flow run 'umber-roadrunner' - Beginning flow run 'umber-roadrunner' for flow 'dummy-flow'
16:34:49.295 | INFO    | Flow run 'umber-roadrunner' - View at http://localhost:4200/runs/flow-run/d0963eac-62bc-4487-9b19-6ebcd8d18dc1
16:34:49.296 | INFO    | prefect.task_runner.dask - Connecting to an existing Dask cluster at 127.0.0.1:8786
16:34:49.305 | INFO    | prefect.task_runner.dask - The Dask dashboard is available at http://127.0.0.1:8787/status
16:35:47.126 | INFO    | Flow run 'umber-roadrunner' - Finished in state Completed()
```

**Dask Worker 1 logs:**

```bash
2025-04-01 16:32:41,935 - distributed.core - INFO - Starting established connection to tcp://127.0.0.1:8786
Task 0
Task 2
Task 4
Task 6
Task 9
Task 7
Task 5
Task 3
Task 1
Task 8
```

**Dask Worker 2 logs:**

```bash
2025-04-01 16:32:43,710 - distributed.core - INFO - Starting established connection to tcp://127.0.0.1:8786
Task 1
Task 3
Task 5
Task 7
Task 8
2025-04-01 16:35:15,535 - distributed._signals - INFO - Received signal SIGINT (2)
2025-04-01 16:35:15,535 - distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:42967. Reason: worker-close
2025-04-01 16:35:15,536 - distributed.nanny - INFO - Closing Nanny at 'tcp://127.0.0.1:40817'. Reason: signal-2
2025-04-01 16:35:15,536 - distributed.nanny - INFO - Nanny asking worker to close. Reason: signal-2
2025-04-01 16:35:15,704 - distributed.dask_worker - INFO - End worke
```

You can see that the tasks `1/3/5/7` are re-ran on Worker 1 when I killed Worker 2 (task `8` was expected to re-run since I killed Worker 2 while running this task).

It's because the [PrefectDaskFuture class](https://github.com/PrefectHQ/prefect/blob/main/src/integrations/prefect-dask/prefect_dask/task_runners.py#L113) holds a reference the `_wrapped_task` (a Dask task) that is never released.

If I update the `PrefectDaskFuture` methods (`wait` and `result`) this way:

```python
class PrefectDaskFuture(PrefectWrappedFuture[R, distributed.Future]):
    def wait(self, timeout: Optional[float] = None) -> None:
        try:
            result = self._wrapped_future.result(timeout=timeout)
            del self._wrapped_future  # <-- Added this
        ...

    def result(
        self,
        timeout: Optional[float] = None,
        raise_on_failure: bool = True,
    ) -> R:
        if not self._final_state:
            try:
                future_result = self._wrapped_future.result(timeout=timeout)
                del self._wrapped_future  # <-- Added this
            ...
```

Then the tasks are not replayed on the workers.

**Comment by cicdw** [comment]
Hi @john-jam thank you for investigating the issue so thoroughly! That makes sense. My only concern would be if removing that reference has consequences for how dask schedules work or resolves dependencies. Could you try out a few other flow structures that have dependencies between tasks and let us know if you see anything problematic? 

If everything seems OK, it would also be great if you wanted to open a PR for this since you identified the solution!
